sensorprocess/sensorprocess.go

```diff
@@ -37,32 +70,64 @@ func (config *Config) StartOfflineSensorProcess(ctx context.Context) bool {
 		return errors.Is(err, replaypcd.ErrEndOfDataset)
 	}
 
-	var imuReading s.TimedIMUReadingResponse
-	if config.IMU != nil {
+	var movementSensorReading s.TimedMovementSensorReadingResponse
+	if config.MovementSensor != nil && (config.MovementSensor.Properties().IMUSupported ||
+		config.MovementSensor.Properties().OdometerSupported) {
 		// get the initial IMU reading; discard all IMU readings that were recorded before the first lidar reading
-		for {
-			movementSensorReading, err := config.IMU.TimedMovementSensorReading(ctx)
-			if err != nil {
-				config.Logger.Warn(err)
-				return errors.Is(err, replaymovementsensor.ErrEndOfDataset)
-			}
-
-			imuReading = *movementSensorReading.TimedIMUResponse
-			if !imuReading.ReadingTime.Before(lidarReading.ReadingTime) {
-				break
-			}
+		movementSensorReading, err = config.getInitialMovementSensorReading(ctx, lidarReading)
+		if err != nil {
+			config.Logger.Warn(err)
+			return errors.Is(err, replaymovementsensor.ErrEndOfDataset)
 		}
 	}
 
+	type readingTime struct {
+		sensorType  string
+		readingTime time.Time
+	}
+
 	// loop over all the data until one of the datasets has reached its end
 	for {
 		select {
 		case <-ctx.Done():
 			return false
 		default:
+			// create a map of supported sensors and their reading time stamps
+			readingTimes := []readingTime{
+				{sensorType: "lidar", readingTime: lidarReading.ReadingTime},
+			}
+			// default to the slightly later imu timestamp, in case that the odometer time stamp was
+			// taken before the lidar time stamp, but the imu time stamp was taken after the lidar time
+			// stamp, we'll want to prioritize adding the lidar measurement before adding the movement
+			// sensor measurement
+			if config.MovementSensor != nil && config.MovementSensor.Properties().IMUSupported {
```

Comment:
[opt] Could always add odom and imu reading times even if they are not supported by properties and then have your sort function put timestamps of time.Time{} at end

Reply:
Don't understand; let's discuss this in person

Reply:
spoke ofline